### PR TITLE
services published are to be executable by everyone

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

- services that are "published" in the "published" templates shall be accessible to everyone in order to use the templates.

This should fix the initial set of access rights.

## Checklist

- [ ] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [ ] Unit tests for the changes exist
- [ ] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: services/catalog/src/simcore_service_catalog/db/repositories/projects.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                & (projects.c.published == True)
+            )
+        ):
+            project_workbench = row.workbench
+            for node in project_workbench:
+                service = project_workbench[node]
+                try:
+                    list_of_published_services.append(ServiceKeyVersion(**service))
+                except ValidationError:
+                    continue
```
Comment (pcrespov) [inline services/catalog/src/simcore_service_catalog/db/repositories/projects.py:26]: log this error to find issues with services metadata

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py [hunk 2]
```diff
@@ -117,59 +130,105 @@ async def _get_service_access_rights(
...
-        except CancelledError:
-            # task is stopped
-            return
-        except Exception:  # pylint: disable=broad-except
-            logger.exception("Error while processing services entry")
-            await asyncio.sleep(
-                5
-            )  # wait a bit before retrying, so it does not block everything until the director is up
+    engine: Engine = app.state.engine
+    async with engine.acquire() as conn:
```
Comment (pcrespov) [inline services/catalog/src/simcore_service_catalog/core/background_tasks.py:212]: This is one of the discussions we had. Having the ``while True`` within the context manager, holds the db connection forever.
why not releasing it before going to sleep?? This way we give a chance to other requests to the database


```python
  while True:
     try:
         async with engine.acquire() as conn:
               _ensure_...(app, conn)
               _ensure_...(app, conn)
         await asyncio.sleep(...)
     except ...
```

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py [hunk 3]
```diff
@@ -117,59 +130,105 @@ async def _get_service_access_rights(
...
+                # check that the list of services is in sync with the registry
+                await _ensure_services_metadata_uptodate(app, conn)
+
+                # check that the published services are available to everyone (templates are published to GUESTs, so their services must be also accessible)
+                await _ensure_published_services_are_accessible(conn)
+
+                await asyncio.sleep(app.state.settings.background_task_rest_time)
+
+            except CancelledError:
+                # task is stopped
```
Comment (pcrespov) [inline services/catalog/src/simcore_service_catalog/core/background_tasks.py:225]: log debug the reason fo the cancelation?

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py [hunk 4]
```diff
@@ -117,59 +130,105 @@ async def _get_service_access_rights(
...
+            "missing services in db:\n%s", pformat(missing_services_in_db),
+        )
+        # update db (rationale: missing services are shared with everyone for now)
+        await _create_services_in_db(
+            connection, missing_services_in_db, services_in_registry
+        )
+
+
+async def _ensure_published_services_are_accessible(connection: SAConnection) -> None:
+    # Rationale: if a project template was published, its services must be available to everyone.
```
Comment (pcrespov) [inline services/catalog/src/simcore_service_catalog/core/background_tasks.py:175]: I would challenge that rationale:  the mechanism should be the other way around. The access-rights to a study is determined by the most restrictive access-right of its components. 

Assume you have a user that creates a study and adds a component with access-rights defined by the component author (e.g. can only be executed by group X). Right now, he has the ability to publish the entire study and disregard the access-rights set by the author of the component.

File: services/catalog/src/simcore_service_catalog/db/repositories/projects.py (hunk 1 above)
Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/db/repositories/projects.py:26]: @pcrespov: it's already done in the background task.

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py (hunk 2 above)
Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/core/background_tasks.py:212]: who are others? just kidding, it's a good point. will do.

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py (hunk 3 above)
Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/core/background_tasks.py:225]: I can add. there is only one possible cancelation in that case, which is when the app stops the background task.

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py (hunk 4 above)
Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/core/background_tasks.py:175]: ok I will rename the function. We are talking here about published templates (the ones accessible from the portal). These ones must have accessible services.

File: services/catalog/src/simcore_service_catalog/core/background_tasks.py (hunk 2 above)
Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/core/background_tasks.py:212]: ok I thought we should acquire once but that makes more sense like this.

**Pull request merged**